src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSource.cs

```diff
@@ -0,0 +1,268 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Core.Imaging;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Adornments;
+using EditorCompletion = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using RoslynTrigger = Microsoft.CodeAnalysis.Completion.CompletionTrigger;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.EditorImplementation
+{
+    internal class CompletionItemSource : IAsyncCompletionSource
+    {
+        internal const string RoslynItem = nameof(RoslynItem);
+        internal const string TriggerBuffer = nameof(TriggerBuffer);
+        internal const string MatchPriority = nameof(MatchPriority);
+        internal const string SelectionBehavior = nameof(SelectionBehavior);
+        internal const string InsertionText = nameof(InsertionText);
+        internal const string UseSuggestionMode = nameof(UseSuggestionMode);
+
+        private static readonly EditorCompletion.CompletionStartData s_emptyCompletionStartData
+            = new EditorCompletion.CompletionStartData(participation: EditorCompletion.CompletionParticipation.DoesNotProvideItems, applicableToSpan: default);
+
+        public EditorCompletion.CompletionStartData InitializeCompletion(EditorCompletion.CompletionTrigger trigger, SnapshotPoint triggerLocation, CancellationToken cancellationToken)
+        {
+            var document = triggerLocation.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return s_emptyCompletionStartData;
+            }
+
+            var service = (CompletionServiceWithProviders)document.GetLanguageService<CompletionService>();
+            if (service == null)
+            {
+                return s_emptyCompletionStartData;
+            }
+
+            var sourceText = document.GetTextSynchronously(cancellationToken);
```

Comment:
sorry, this wasn't explained.  Why isn't this async?

Reply:
The current method should be call synchronously. This is the first call from editor to us. We should decide if we want to provide items to completion or not. 

In the next call (GetCompletionContextAsync), we can use async methods but here there is no reason to do so.

---
In reply to: [222078854](https://github.com/dotnet/roslyn/pull/29016#discussion_r222078854) [](ancestors = 222078854)

Reply:
As our own documentation says:

```c#
        /// <remarks>
        /// This API uses SourceText instead of Document so implementations can only be based on text, not syntax or semantics.
        /// </remarks>
```

We should not be going to a document here, and we definitely should not be executing potentially expensive stuff like GetTextSynchronously.  Since we need the SourceText, we shoudl just be doing: `triggerLocation.Snapshot.AsText()`.